### PR TITLE
[BugFix] fix swap mv and alter view of mv

### PR DESCRIPTION
**Pull request by murphyatwork**
Fixes #issue

Several bugfixs:
1. swap table: should recursive inactive related MVs 
2. alter view: should inactive the related MV

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by packy92 — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 29 / 34 (85.29%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/CreateMaterializedViewStmt.java|1|2|50.00%|[624]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|4|6|66.67%|[3268, 3286]|
|:large_blue_circle:|com/starrocks/alter/AlterJobMgr.java|5|6|83.33%|[1039]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|12|13|92.31%|[672]|
|:large_blue_circle:|com/starrocks/catalog/BaseTableInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 29 / 33 (87.88%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|4|6|66.67%|[3268, 3286]|
|:large_blue_circle:|com/starrocks/alter/AlterJobMgr.java|5|6|83.33%|[1039]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|13|14|92.86%|[672]|
|:large_blue_circle:|com/starrocks/catalog/BaseTableInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 29 / 33 (87.88%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|4|6|66.67%|[3268, 3286]|
|:large_blue_circle:|com/starrocks/alter/AlterJobMgr.java|5|6|83.33%|[1039]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|13|14|92.86%|[672]|
|:large_blue_circle:|com/starrocks/catalog/BaseTableInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 30 / 33 (90.91%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|5|6|83.33%|[3286]|
|:large_blue_circle:|com/starrocks/alter/AlterJobMgr.java|5|6|83.33%|[1039]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|13|14|92.86%|[672]|
|:large_blue_circle:|com/starrocks/catalog/BaseTableInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 30 / 33 (90.91%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|5|6|83.33%|[3286]|
|:large_blue_circle:|com/starrocks/alter/AlterJobMgr.java|5|6|83.33%|[1039]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|13|14|92.86%|[672]|
|:large_blue_circle:|com/starrocks/catalog/BaseTableInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 30 / 33 (90.91%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|5|6|83.33%|[3286]|
|:large_blue_circle:|com/starrocks/alter/AlterJobMgr.java|5|6|83.33%|[1039]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|13|14|92.86%|[672]|
|:large_blue_circle:|com/starrocks/catalog/BaseTableInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|1|1|100.00%|[]|

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 29 / 33 (87.88%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|4|6|66.67%|[3272, 3290]|
|:large_blue_circle:|com/starrocks/alter/AlterJobMgr.java|5|6|83.33%|[1039]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|13|14|92.86%|[672]|
|:large_blue_circle:|com/starrocks/catalog/BaseTableInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 29 / 33 (87.88%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|4|6|66.67%|[3272, 3290]|
|:large_blue_circle:|com/starrocks/alter/AlterJobMgr.java|5|6|83.33%|[1039]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|13|14|92.86%|[672]|
|:large_blue_circle:|com/starrocks/catalog/BaseTableInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|1|1|100.00%|[]|

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 29 / 33 (87.88%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|4|6|66.67%|[3272, 3290]|
|:large_blue_circle:|com/starrocks/alter/AlterJobMgr.java|5|6|83.33%|[1039]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|13|14|92.86%|[672]|
|:large_blue_circle:|com/starrocks/catalog/BaseTableInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|1|1|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java [hunk 1]
```diff
@@ -198,7 +188,11 @@ private static void processViews(QueryStatement queryStatement, List<BaseTableIn
         }
         Set<ViewRelation> viewRelationSet = Sets.newHashSet(viewRelations);
         for (ViewRelation viewRelation : viewRelationSet) {
+            // base tables of view
             processBaseTables(viewRelation.getQueryStatement(), baseTableInfos);
+
+            // view itself is considered as base-table
+            baseTableInfos.add(BaseTableInfo.fromTableName(viewRelation.getName(), viewRelation.getView()));
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java:195]: Why add view as the base table? To keep the relation between MV and view?

File: fe/fe-core/src/main/java/com/starrocks/alter/AlterJobMgr.java [hunk 2]
```diff
@@ -998,6 +1005,7 @@ public void processAlterView(AlterViewStmt stmt, ConnectContext ctx) throws User
             }
             view.setNewFullSchema(newFullSchema);
 
+            LocalMetastore.inactiveRelatedMaterializedView(db, view, String.format("base view %s changed", viewName));
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/alter/AlterJobMgr.java:1008]: Can we active the mv later after view has changed ?

**Review by Astralidea — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java (hunk 1 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java:195]: To make the MV be aware of view changing, specifically you alter the view, the MV will be inactivated

File: fe/fe-core/src/main/java/com/starrocks/alter/AlterJobMgr.java (hunk 2 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/alter/AlterJobMgr.java:1008]: yes

**Review by satanson — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by mofeiatwork** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**